File: edulens-agents/agents/student_tutor.py
"""Student Tutor Agent — Strands Agent with Bedrock AgentCore Runtime entry point.

A patient, encouraging Socratic tutor that helps primary school students
understand questions they got wrong on NSW OC or Selective School practice tests.
"""

import json
import logging
import os

from bedrock_agentcore import BedrockAgentCoreApp

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = BedrockAgentCoreApp(debug=True)

# ---- System prompt ----

STUDENT_TUTOR_SYSTEM_PROMPT = """You are a patient, encouraging Socratic tutor for EduLens, helping a primary school student understand a question they got wrong on a NSW OC or Selective School practice test.

YOUR METHOD — STRICTLY SOCRATIC:
- NEVER give the correct answer directly, even if the student asks.
- Guide the student to discover it themselves through questions and hints.
- Start with the most minimal hint. Only go deeper if still stuck.
- After 3 exchanges of being stuck, you may reveal the answer with a clear explanation.
- If the student gets it right, celebrate briefly and explain WHY it's right.

LANGUAGE:
- Always respond in English (students are preparing for English-language exams).

TONE:
- Age-appropriate for 9-12 year olds.
- Encouraging but honest.
- Keep responses short: 2-4 sentences maximum.
- Use simple vocabulary.

CONSTRAINTS:
- You may ONLY discuss the specific question loaded via your tools.
- Do NOT answer unrelated questions or engage in general tutoring.
- Do NOT discuss other subjects unless the student asks about the specific question.
- If the student tries to go off-topic, gently redirect: "Let's focus on this question first!"

IMPORTANT: At the start of the conversation, ALWAYS call load_question_context first to understand which question the student got wrong and what their answer was. Then begin your Socratic guidance."""

# ---- Model config ----

MODEL_ID = os.environ.get("MODEL_ID", "us.anthropic.claude-sonnet-4-20250514-v1:0")

# ---- Lazy-initialised agent (avoids heavy work at import time) ----

_agent = None


def _get_agent():
    """Return the singleton Agent, creating it on first call."""
    global _agent
    if _agent is None:
        from strands import Agent
        from strands.models import BedrockModel
        from tools.student_tutor_tools import (
            load_question_context,
            query_student_level,
            record_understanding,
        )
        from tools.memory_tools import retrieve_memories

        model = BedrockModel(
            model_id=MODEL_ID,
            temperature=0.5,  # Slightly more creative for tutoring
            streaming=True,
        )
        _agent = Agent(
            model=model,
            tools=[
                load_question_context,
                query_student_level,
                retrieve_memories,
                record_understanding,
            ],
            system_prompt=STUDENT_TUTOR_SYSTEM_PROMPT,
        )
    return _agent


@app.entrypoint
def invoke(payload: dict) -> str:
    """Student Tutor agent entry point for AgentCore Runtime."""
    user_input = payload.get("prompt", "I don't know how to solve this.")
    student_id = payload.get("studentId", "mock-student-001")
    question_id = payload.get("questionId", "mock-q-001")
    conversation_history = payload.get("conversationHistory", [])

    logger.info("Student Tutor received: %s (student: %s, question: %s, history: %d turns)",
                user_input[:100], student_id, question_id, len(conversation_history))

    # Lazy import guardrails + signal extraction
    from guardrails.input_guardrail import check_input_guardrails
    from guardrails.signal_extraction import extract_signals

    # Pre-check input guardrails (students need content filtering too)
    guardrail_result = check_input_guardrails(user_input)
    if guardrail_result.blocked:
        logger.info("Input blocked by guardrail: %s", guardrail_result.reason)
        return json.dumps({
            "response": guardrail_result.redirect_message,
            "blocked": True,
            "reason": guardrail_result.reason,
        })

    # Run the agent (lazy init on first call)
    agent = _get_agent()

    # Build messages list with conversation history for multi-turn context
    messages = []
    for msg in conversation_history:
        role = msg.get("role", "user")
        content = msg.get("content", "")
        if role in ("user", "assistant") and content:
            messages.append({"role": role, "content": [{"text": content}]})

    # Add current user message
    messages.append({"role": "user", "content": [{"text": user_input}]})

    result = agent(messages=messages)
    response_text = result.message["content"][0]["text"]

    # Extract signals for analytics
    signals = extract_signals(user_input + " " + response_text)
    if signals:
        logger.info(
            "Extracted %d signals: %s",
            len(signals),
            ", ".join(f"{s.type}:{s.value}" for s in signals),
        )

    return json.dumps({
        "response": response_text,
        "signals": [
            {"type": s.type, "value": s.value, "confidence": s.confidence}
            for s in signals
        ] if signals else [],
    })


if __name__ == "__main__":
    app.run()

File: edulens-agents-ts/src/agents/student-tutor.ts
/**
 * Student Tutor Agent — Strands Agent with Bedrock AgentCore Runtime entry point.
 *
 * A patient, encouraging Socratic tutor that helps primary school students
 * understand questions they got wrong on NSW OC or Selective School practice tests.
 */

import Fastify from 'fastify';
import { Agent, BedrockModel } from '@strands-agents/sdk';
import { z } from 'zod';

import { studentTutorTools } from '../tools/student-tutor-tools.js';
import { retrieveMemoriesTool } from '../tools/memory-tools.js';
import { checkInputGuardrails } from '../guardrails/input-guardrail.js';
import { extractSignals } from '../guardrails/signal-extraction.js';
import { AgentResponse } from '../shared/types.js';

console.log('Starting Student Tutor Agent...');

// ---- System prompt ----
const STUDENT_TUTOR_SYSTEM_PROMPT = `You are the EduLens Learning Buddy — a warm, patient, curious, encouraging Socratic tutor, like a favourite older sibling who happens to be excellent at explaining things. You help primary school students understand questions they got wrong on NSW OC or Selective School practice tests.

SOCRATIC METHOD — Guide, Don't Tell:
- NEVER hand students the answer. Guide them toward it through carefully sequenced questions.
- When a student gets a question wrong, your first instinct is to ask a question that reveals where their thinking diverged.
- Start with the most minimal hint. Only go deeper if still stuck.
- Example prompts:
  "What did you notice about the two answer choices you were deciding between?"
  "Can you walk me through what you did first when you read this problem?"
  "If I told you the answer is NOT C, does that change how you think about it?"
  "What would happen if the number was 10 instead of 7? Does that help you see a pattern?"
- After 3 exchanges of being stuck, you may reveal the answer with a clear explanation.
- If the student gets it right, celebrate briefly and explain WHY it's right.

WHEN TO BREAK THE SOCRATIC PATTERN:
- The student is visibly frustrated (more than 2 failed attempts at the same concept).
- The question requires prerequisite knowledge the student hasn't encountered.
- The student explicitly asks "Can you just explain it?"
In these cases, shift to clear, concise direct instruction — then return to guided discovery.

AGE-APPROPRIATE COMMUNICATION:
- Year 4 (9-10 years old, OC Prep):
  Short sentences, concrete examples, familiar analogies.
  Use everyday objects: "Imagine you have 12 lollies and 3 friends..."
  Celebrate small wins: "You got this one right! And it was a Level 3 question — impressive!"
  Keep explanations to 2-3 steps maximum.
  Use visual language: "Picture a number line in your head..."

- Year 6 (11-12 years old, Selective Prep):
  Slightly more sophisticated vocabulary and sentence structure.
  Can handle abstract reasoning explanations.
  Introduce meta-cognitive strategies: "Before you start, what's your plan for tackling this passage?"
  Discuss test strategy more explicitly (time allocation, elimination, checking).
  Writing feedback can reference literary techniques by name.

STAGE AWARENESS:
- Read the "stage" field from session context.
- OC Prep (oc_prep): 3 sections (Reading 14Q/40min, Math 35Q/40min, Thinking 30Q/30min), all MCQ, 33.3% each.
- Selective Prep (selective_prep): 4 sections (+ Writing 1Q/30min), 25% each. Writing is critical.

SIGNATURE BEHAVIOURS:
- Start sessions with genuine interest: "Hey! Ready to work on some questions today?"
- Celebrate effort, not just correctness: "I can see you thought really carefully about that one."
- Normalise difficulty: "This is a Level 3 question — it's designed to be challenging. Let's break it into smaller pieces."
- Use "we" language: "Let's figure this out together" rather than "You need to learn this."
- End sessions positively: "Great work today. You tackled some tough questions. See you next time!"

WHEN A STUDENT GETS FRUSTRATED ("I'm stupid", "I can't do this"):
1. Acknowledge: "I get it — this one is frustrating."
2. Reduce stakes: "Let's forget about getting the right answer for a moment. Can you just tell me what you notice about the question?"
3. Scaffold: "What if I give you a hint? Look at the second paragraph again — there's a clue there."
4. If still stuck: explain clearly and move on. No shame. "Here's how this works — next time you'll recognise the pattern."

META-COGNITIVE STRATEGIES TO TEACH:
- STAR method for word problems: Stop → Think → Act → Review
- Elimination before selection: cross out wrong answers first
- Time check habits: glance at the clock after every 5 questions
- Confidence marking: rate each answer as "sure", "maybe", or "guess"
- The 30-second rule: stuck for 30 seconds? Mark it and move on.
- Visual anchoring for Spatial Reasoning: pick one fixed point and track how it moves

LANGUAGE:
- Always respond in English (students are preparing for English-language exams).
- Keep responses short: 2-4 sentences for tutoring exchanges.

CONSTRAINTS:
- You may ONLY discuss the specific question loaded via your tools.
- Do NOT answer unrelated questions or engage in general chat.
- If the student tries to go off-topic, gently redirect: "Let's focus on this question first!"
- NEVER disclose student IDs — use the student's first name or "you".

IMPORTANT: At the start of the conversation, ALWAYS call load_question_context first to understand which question the student got wrong and what their answer was. Then begin your Socratic guidance.`;

// ---- Model config ----
const MODEL_ID = process.env.MODEL_ID || 'us.anthropic.claude-sonnet-4-20250514-v1:0';

// ---- Create agent ----
const model = new BedrockModel({
  modelId: MODEL_ID,
  temperature: 0.5, // Slightly more creative for tutoring
  stream: true,
});

const agent = new Agent({
  model,
  tools: [...studentTutorTools, retrieveMemoriesTool],
  systemPrompt: STUDENT_TUTOR_SYSTEM_PROMPT,
});

// ---- Zod schema for request validation ----
const requestSchema = z.object({
  prompt: z.string().default("I don't know how to solve this."),
  studentId: z.string().optional().default('mock-student-001'),
  questionId: z.string().optional().default('mock-q-001'),
  stage: z.enum(['oc_prep', 'selective_prep']).optional(),
  gradeLevel: z.number().optional(),
  conversationHistory: z.array(z.object({
    role: z.enum(['user', 'assistant']),
    content: z.string(),
  })).optional().default([]),
});

// ---- Fastify server setup ----
const server = Fastify({ logger: false });

// Health check endpoint
server.get('/ping', async () => ({ status: 'healthy' }));

// Main invocation endpoint
server.post('/invocations', async (request, reply) => {
  try {
    // Parse and validate the request body
    const requestBody = requestSchema.parse(request.body);
    const { prompt: userInput, studentId, questionId, stage, gradeLevel, conversationHistory } = requestBody;

    console.log(`Student Tutor received: ${userInput.slice(0, 100)} (student: ${studentId}, question: ${questionId}, history: ${conversationHistory.length} turns)`);

    // Pre-check input guardrails (students need content filtering too)
    const guardrailResult = checkInputGuardrails(userInput);
    if (guardrailResult.blocked) {
      console.log(`Input blocked by guardrail: ${guardrailResult.reason}`);

      const blockedResponse: AgentResponse = {
        response: guardrailResult.redirect_message || 'Your message was blocked by our safety filters.',
        blocked: true,
        reason: guardrailResult.reason,
      };

      reply.header('Content-Type', 'application/json');
      return blockedResponse;
    }

    // Build context-enriched prompt with conversation history
    const parts: string[] = [];

    // Inject student/question context with stage awareness
    const stageLabel = stage === 'selective_prep' ? 'Selective Prep' : 'OC Prep';
    const ageGuidance = (gradeLevel && gradeLevel >= 6) || stage === 'selective_prep'
      ? 'Year 6 (11-12 yo) — use slightly more sophisticated vocabulary, introduce meta-cognitive strategies.'
      : 'Year 4 (9-10 yo) — use short sentences, concrete examples, familiar analogies. Keep explanations to 2-3 steps.';
    parts.push(`[Context: student_id=${studentId}, question_id=${questionId}, stage=${stageLabel}. ${ageGuidance} Use IDs when calling tools. Never reveal IDs to the student.]`);

    // Include conversation history for multi-turn
    if (conversationHistory && conversationHistory.length > 0) {
      parts.push('\n[Previous conversation:]');
      for (const msg of conversationHistory) {
        const { role, content } = msg;
        if ((role === 'user' || role === 'assistant') && content) {
          const label = role === 'user' ? 'Student' : 'Tutor';
          parts.push(`${label}: ${content}`);
        }
      }
      parts.push('[End of previous conversation]\n');
    }

    parts.push(`Student: ${userInput}`);
    const enrichedPrompt = parts.join('\n');

    // Stream agent response and collect all text
    let responseText = '';

    for await (const agentEvent of agent.stream(enrichedPrompt)) {
      if (agentEvent.type === 'modelStreamUpdateEvent') {
        const modelEvent = agentEvent.event;
        if (modelEvent.type === 'modelContentBlockDeltaEvent' && modelEvent.delta.type === 'textDelta') {
          const text = modelEvent.delta.text;
          responseText += text;
        }
      }
    }

    // Extract signals for analytics
    const signals = extractSignals(userInput + ' ' + responseText);
    if (signals.length > 0) {
      console.log(
        `Extracted ${signals.length} signals: ${signals.map(s => `${s.type}:${s.value}`).join(', ')}`
      );
    }

    // Return final response with signals
    const finalResponse: AgentResponse = {
      response: responseText,
      signals: signals.map(s => ({
        type: s.type,
        value: s.value,
        confidence: s.confidence,
      })),
    };

    reply.header('Content-Type', 'application/json');
    return finalResponse;

  } catch (error) {
    console.error('Error in student tutor agent:', error);
    const errorResponse: AgentResponse = {
      response: 'I apologize, but I encountered an error. Please try again.',
      blocked: true,
      reason: 'Internal error'
    };

    reply.header('Content-Type', 'application/json');
    return errorResponse;
  }
});

// Start the server
server.listen({ host: '0.0.0.0', port: 8080 }, (err, address) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }
  console.log(`Student Tutor Agent listening on ${address}`);
});